Test on AppVeyor + Fix the regression around < on Windows

eed3si9n opened the pull request:
Ref https://github.com/sbt/zinc/issues/757


Prior to https://github.com/sbt/zinc/issues/712, this code looked like:

```scala
          // This lookup could be improved if a hint where to look is given.
          outputDirs.map(new File(_, classFilePath)).find(_.exists()).map(AbstractFile.getFile(_))
```

When I ported this to NIO Path, it started to error on Windows because `<` are not allowed as path character. This change should at least get back to the state, hopefully.

eed3si9n committed:
Test on AppVeyor

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=develop&zinc=pull/759/head&io=develop&librarymanagement=develop&util=develop&website=develop&deploy_to_bintray=false&run_sbt_community=false&verbose_dbuild=false&community_verbose=false">this link.</A>

eed3si9n commented:
https://github.com/scalapb/protoc-bridge/issues/56
https://ci.appveyor.com/project/sbt/zinc/builds/32372226

```
[error] java.lang.UnsupportedClassVersionError: protocbridge/frontend/BridgeApp
[error] 	at protocbridge.frontend.WindowsPluginFrontend$.createWindowsScript(WindowsPluginFrontend.scala:47)
[error] 	at protocbridge.frontend.WindowsPluginFrontend$.prepare(WindowsPluginFrontend.scala:21)
[error] 	at protocbridge.ProtocBridge$.$anonfun$runWithGenerators$1(ProtocBridge.scala:117)
[error] 	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:238)
[error] 	at scala.collection.immutable.List.foreach(List.scala:392)
[error] 	at scala.collection.TraversableLike.map(TraversableLike.scala:238)
[error] 	at scala.collection.TraversableLike.map$(TraversableLike.scala:231)
[error] 	at scala.collection.immutable.List.map(List.scala:298)
[error] 	at protocbridge.ProtocBridge$.runWithGenerators(ProtocBridge.scala:116)
[error] 	at protocbridge.ProtocBridge$.run(ProtocBridge.scala:81)
[error] 	at sbtprotoc.ProtocPlugin$.executeProtoc(ProtocPlugin.scala:172)
[error] 	at sbtprotoc.ProtocPlugin$.compile(ProtocPlugin.scala:209)
[error] 	at sbtprotoc.ProtocPlugin$.compileProto$1(ProtocPlugin.scala:294)
[error] 	at sbtprotoc.ProtocPlugin$.$anonfun$sourceGeneratorTask$7(ProtocPlugin.scala:302)
```

eed3si9n committed:
ScalaPB 0.10.2

eed3si9n commented:
```
[info] JavaCompilerSpec:
[info] Compiling a java file with local javac
[info] - should compile a java file *** FAILED ***
[info]   java.net.URISyntaxException: Illegal character in path at index 10: vf:/tmp/C:\projects\zinc\internal\zinc-compile-core\target\jvm-2.12\test-classes\sbt\internal\inc\javac\good.java
[info]   at java.net.URI$Parser.fail(URI.java:2848)
[info]   at java.net.URI$Parser.checkChars(URI.java:3021)
[info]   at java.net.URI$Parser.parseHierarchical(URI.java:3105)
[info]   at java.net.URI$Parser.parse(URI.java:3053)
[info]   at java.net.URI.<init>(URI.java:588)
[info]   at sbt.internal.inc.javac.LocalJava$.toUri(LocalJava.scala:196)
[info]   at sbt.internal.inc.javac.LocalJava$.toFileObject(LocalJava.scala:173)
[info]   at sbt.internal.inc.javac.LocalJavaCompiler.$anonfun$run$8(LocalJava.scala:312)
[info]   at scala.collection.immutable.List.map(List.scala:286)
[info]   at sbt.internal.inc.javac.LocalJavaCompiler.run(LocalJava.scala:312)
```

eed3si9n committed:
Change backslash to forward slash on PlainVirtualFile id

eed3si9n committed:
Fix the Windows regression around <

Prior to zinc 712, this code looked like:

```scala
          // This lookup could be improved if a hint where to look is given.
          outputDirs.map(new File(_, classFilePath)).find(_.exists()).map(AbstractFile.getFile(_))
```

When I ported this to NIO Path, it started to error on Windows because `<` are not allowed as path character. This change should at least get back to the state, hopefully.

eed3si9n commented:
https://ci.appveyor.com/project/sbt/zinc/builds/32373212#L2682

```
[info] CachedHashingSpec:
[info] zinc
[info] - should cache jar generation *** FAILED ***
[info]   67 was not less than 42.0 Cache jar didn't work: 67 is >= than 50% of 84. (CachedHashingSpec.scala:79)
```

eed3si9n committed:
Skip CachedHashingSpec on Windows

eed3si9n commented:
https://ci.appveyor.com/project/sbt/zinc/builds/32373674

```scala
[error] x source-dependencies / canon
[error] sbt.internal.inc.BatchScriptRunner$PreciseScriptedError: Command failed: 'compile'
[error] Caused by: scala.reflect.internal.FatalError: Error accessing C:\Users\appveyor\AppData\Local\Temp\1\sbt_e421fae9\lib\a.jar
[error] 	at scala.tools.nsc.classpath.AggregateClassPath.$anonfun$list$3(AggregateClassPath.scala:99)
[error] 	at scala.collection.Iterator.foreach(Iterator.scala:941)
[error] 	at scala.collection.Iterator.foreach$(Iterator.scala:941)
[error] 	at scala.collection.AbstractIterator.foreach(Iterator.scala:1429)
[error] 	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
[error] 	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
[error] 	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
[error] 	at scala.tools.nsc.classpath.AggregateClassPath.list(AggregateClassPath.scala:87)
[error] 	at scala.tools.nsc.util.ClassPath.list(ClassPath.scala:36)
[error] 	at scala.tools.nsc.util.ClassPath.list$(ClassPath.scala:36)
[error] 	at scala.tools.nsc.classpath.AggregateClassPath.list(AggregateClassPath.scala:30)
[error] 	at scala.tools.nsc.symtab.SymbolLoaders$PackageLoader.doComplete(SymbolLoaders.scala:284)
[error] 	at scala.tools.nsc.symtab.SymbolLoaders$SymbolLoader.complete(SymbolLoaders.scala:230)
[error] 	at scala.reflect.internal.Symbols$Symbol.info(Symbols.scala:1542)
[error] 	at scala.reflect.internal.Mirrors$RootsBase.init(Mirrors.scala:257)
[error] 	at scala.tools.nsc.Global.rootMirror$lzycompute(Global.scala:74)
[error] 	at scala.tools.nsc.Global.rootMirror(Global.scala:72)
[error] 	at scala.tools.nsc.Global.rootMirror(Global.scala:44)
[error] 	at scala.reflect.internal.Definitions$DefinitionsClass.ObjectClass$lzycompute(Definitions.scala:295)
[error] 	at scala.reflect.internal.Definitions$DefinitionsClass.ObjectClass(Definitions.scala:295)
[error] 	at scala.reflect.internal.Definitions$DefinitionsClass.init(Definitions.scala:1480)
[error] 	at scala.tools.nsc.Global$Run.<init>(Global.scala:1199)
[error] 	at xsbt.ZincCompiler$ZincRun.<init>(CallbackGlobal.scala:79)
[error] 	at xsbt.CachedCompiler0.run(CompilerInterface.scala:155)
[error] 	at xsbt.CachedCompiler0.run(CompilerInterface.scala:129)
[error] 	at xsbt.CompilerInterface.run(CompilerInterface.scala:40)
[error] 	at sun.reflect.GeneratedMethodAccessor43.invoke(Unknown Source)
```

eed3si9n commented:
`source-dependencies / canon` uses symlink, so the failure actually makes sense.

eed3si9n committed:
Disable source-dependencies/canon

This test will fail on Windows.

retronym commented on internal/compiler-bridge/src/main/scala/xsbt/CallbackGlobal.scala:
```diff
@@ -171,10 +171,12 @@ sealed class ZincCompiler(settings: Settings, dreporter: DelegatingReporter, out
-            .map(_.resolve(classFilePath))
-            .find(Files.exists(_))
-            .map(Compat.plainNioFile(_))
+          if (classFilePath.contains("<")) None
```
Okay as a defensive backstop, but we should require the inputs to this method to be valid classfile paths. My fix to `fullName` and `processDependency` _should_ get us there.

eed3si9n replied:
Yea. For this PR I just wanted to get AppVeyor going, but I'd be happy to remove this part if/when it's no longer needed.

retronym commented on internal/zinc-compile-core/src/main/scala/sbt/internal/inc/PlainVirtualFile.scala:
```diff
@@ -17,7 +17,9 @@ import java.io.InputStream
 
-class PlainVirtualFile(path: Path) extends BasicVirtualFileRef(path.toString) with PathBasedFile {
+class PlainVirtualFile(path: Path)
+    extends BasicVirtualFileRef(path.toString.replaceAllLiterally("\\", "/"))
```
Nitpick, but I'd prefer using `replace('\\', '/')`  (a method on `String`) here.

eed3si9n replied:
done - https://github.com/sbt/zinc/pull/759/commits/532920dd3aaffb0355969b366c5cd41151b84759

retronym commented:
The test failure on appveyor:

```
[error] (run-main-0) sbt.internal.inc.ScriptedRunnerImpl$ScriptedFailure$1: 1 (out of 139) scripted tests failed:
[error] 
[error] 	source-dependencies / sealed
[error] sbt.internal.inc.ScriptedRunnerImpl$ScriptedFailure$1: 1 (out of 139) scripted tests failed:
[error] 
[error] 	source-dependencies / sealed
[error] stack trace is suppressed; run 'last zincScriptedJVM2_12 / Test / bgRun' for the full output
[error] Nonzero exit code: 1
[error] (zincScriptedJVM2_12 / Test / run) Nonzero exit code: 1
[error] Total time: 137 s (02:17), completed Apr 22, 2020 11:17:35 PM
```

Isn't SBT supposed to unsuppress stack traces automatically when running in without a terminal attached (e.g. running in CI)?

eed3si9n commented:
> Isn't SBT supposed to unsuppress stack traces automatically when running in without a terminal attached (e.g. running in CI)?

For a while it was showing stack traces all the time so I reimplemented the suppression at some point, maybe I didn't translate the subtle unsuppress behavior.

eed3si9n commented:
So this looks like `java.nio.file.Files.copy` failed in the test code. 

<details>
<summary>

```scala
[error] x source-dependencies / sealed
[error] sbt.internal.inc.BatchScriptRunner$PreciseScriptedError: Command failed: 'compile'
[error] Caused by: java.nio.file.FileSystemException: C:\Users\appveyor\AppData\Local\Temp\1\sbt_320b713f\target-bridge-2.12.10.jar: The process cannot access the file because it is being used by another process.
```

</summary>


```scala
[error] x source-dependencies / sealed
[error] sbt.internal.inc.BatchScriptRunner$PreciseScriptedError: Command failed: 'compile'
[error] Caused by: java.nio.file.FileSystemException: C:\Users\appveyor\AppData\Local\Temp\1\sbt_320b713f\target-bridge-2.12.10.jar: The process cannot access the file because it is being used by another process.
[error] 
[error] 	at sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:86)
[error] 	at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:97)
[error] 	at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:102)
[error] 	at sun.nio.fs.WindowsFileCopy.copy(WindowsFileCopy.java:165)
[error] 	at sun.nio.fs.WindowsFileSystemProvider.copy(WindowsFileSystemProvider.java:278)
[error] 	at java.nio.file.Files.copy(Files.java:1274)
[error] 	at sbt.internal.inc.AbstractBridgeProviderTestkit.getCompilerBridge(AbstractBridgeProviderTestkit.scala:28)
[error] 	at sbt.internal.inc.AbstractBridgeProviderTestkit.getCompilerBridge$(AbstractBridgeProviderTestkit.scala:21)
[error] 	at sbt.internal.inc.BridgeProviderSpecification.getCompilerBridge(BridgeProviderSpecification.scala:19)
[error] 	at sbt.internal.inc.IncHandler.onNewIncInstance(IncHandler.scala:163)
[error] 	at sbt.internal.inc.IncHandler.$anonfun$onIncInstance$1(IncHandler.scala:151)
[error] 	at scala.Option.getOrElse(Option.scala:189)
[error] 	at sbt.internal.inc.IncHandler.onIncInstance(IncHandler.scala:151)
[error] 	at sbt.internal.inc.IncHandler.apply(IncHandler.scala:145)
[error] 	at sbt.internal.inc.IncHandler.apply(IncHandler.scala:63)
[error] 	at sbt.internal.inc.BatchScriptRunner.processStatement(BatchScriptRunner.scala:45)
[error] 	at sbt.internal.inc.BatchScriptRunner.$anonfun$apply$1(BatchScriptRunner.scala:27)
[error] 	at sbt.internal.inc.BatchScriptRunner.$anonfun$apply$1$adapted(BatchScriptRunner.scala:27)
[error] 	at scala.collection.immutable.List.foreach(List.scala:392)
[error] 	at sbt.internal.inc.BatchScriptRunner.apply(BatchScriptRunner.scala:27)
[error] 	at sbt.internal.inc.ScriptedTests.$anonfun$commonRunTest$8(ScriptedTests.scala:234)
[error] 	at scala.util.control.Exception$Catch.apply(Exception.scala:228)
[error] 	at sbt.internal.inc.ScriptedTests.commonRunTest(ScriptedTests.scala:231)
[error] 	at sbt.internal.inc.ScriptedTests.$anonfun$runBatchedTests$2(ScriptedTests.scala:163)
[error] 	at sbt.internal.inc.ScriptedTests.runOrHandleDisabled(ScriptedTests.scala:181)
[error] 	at sbt.internal.inc.ScriptedTests.$anonfun$runBatchedTests$1(ScriptedTests.scala:164)
[error] 	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:238)
[error] 	at scala.collection.mutable.ArraySeq.foreach(ArraySeq.scala:75)
[error] 	at scala.collection.TraversableLike.map(TraversableLike.scala:238)
[error] 	at scala.collection.TraversableLike.map$(TraversableLike.scala:231)
[error] 	at scala.collection.AbstractTraversable.map(Traversable.scala:108)
[error] 	at sbt.internal.inc.ScriptedTests.runBatchTests$1(ScriptedTests.scala:147)
[error] 	at sbt.internal.inc.ScriptedTests.runBatchedTests(ScriptedTests.scala:170)
[error] 	at sbt.internal.inc.ScriptedTests.$anonfun$batchScriptedRunner$10(ScriptedTests.scala:82)
[error] 	at sbt.io.IO$.withTemporaryDirectory(IO.scala:489)
[error] 	at sbt.io.IO$.withTemporaryDirectory(IO.scala:499)
[error] 	at sbt.internal.inc.ScriptedTests.$anonfun$batchScriptedRunner$9(ScriptedTests.scala:81)
[error] 	at sbt.internal.inc.ScriptedRunnerImpl$.$anonfun$runAllInParallel$1(IncScriptedRunner.scala:66)
[error] 	at scala.collection.parallel.mutable.ParArray$ParArrayIterator.flatmap2combiner(ParArray.scala:419)
[error] 	at scala.collection.parallel.ParIterableLike$FlatMap.leaf(ParIterableLike.scala:1074)
[error] 	at scala.collection.parallel.Task.$anonfun$tryLeaf$1(Tasks.scala:53)
[error] 	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] 	at scala.util.control.Breaks$$anon$1.catchBreak(Breaks.scala:67)
[error] 	at scala.collection.parallel.Task.tryLeaf(Tasks.scala:56)
[error] 	at scala.collection.parallel.Task.tryLeaf$(Tasks.scala:50)
[error] 	at scala.collection.parallel.ParIterableLike$FlatMap.tryLeaf(ParIterableLike.scala:1070)
[error] 	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.internal(Tasks.scala:160)
[error] 	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.internal$(Tasks.scala:157)
[error] 	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.internal(Tasks.scala:440)
[error] 	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.compute(Tasks.scala:150)
[error] 	at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask.compute$(Tasks.scala:149)
[error] 	at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.compute(Tasks.scala:440)
[error] 	at java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:189)
[error] 	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
[error] 	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
[error] 	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
[error] 	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

</details>
